chain/events/filter/index.go

```diff
@@ -317,11 +322,111 @@ func (ei *EventIndex) migrateToVersion3(ctx context.Context) error {
 	if err != nil {
 		return xerrors.Errorf("commit transaction: %w", err)
 	}
-	log.Infof("Successfully migrated events to version 3 in %s", time.Since(now))
+	log.Infof("successfully migrated event index from version 2 to version 3 in %s", time.Since(now))
+	return nil
+}
+
+// migrateToVersion4 migrates the schema from version 3 to version 4: indexing events by emitter actor ID.
+// This migration replaces the emitter_addr column in event table with a new column called `emitter`, which stores
+// the emitter's actor ID.
+// The migration
+// For each event
+func (ei *EventIndex) migrateToVersion4(ctx context.Context, chainStore *store.ChainStore, actorResolver ActorResolver) error {
+	now := time.Now()
+
+	head := chainStore.GetHeaviestTipSet()
+
+	tx, err := ei.db.BeginTx(ctx, nil)
+	if err != nil {
+		return xerrors.Errorf("begin transaction: %w", err)
+	}
+	// Rollback the transaction (a no-op if the transaction was already committed)
+	defer func() { _ = tx.Rollback() }()
+
+	// Alter the event table to add a new column called `emitter`
+	// Note: since `emitter` column does not accept NULL values, set the default to 0 so that the table
+	// can be altered. This means after the migration all the reverted events for which address resolution
+	// cannot be performed will end up with 0 as their emitter actor ID.
+	if _, err = tx.Exec("ALTER TABLE event ADD COLUMN emitter INTEGER NOT NULL DEFAULT 0"); err != nil {
+		return xerrors.Errorf("increment _meta version: %w", err)
+	}
+
+	stmtUpdateEmitterByID, err := tx.Prepare("UPDATE event SET emitter=? WHERE id=?")
+	if err != nil {
+		return xerrors.Errorf("prepare stmtUpdateEmitterByID: %w", err)
+	}
+
+	rows, err := tx.QueryContext(ctx, `SELECT id, emitter_addr FROM event WHERE reverted=false ORDER BY id DESC`)
+	if err != nil {
+		return xerrors.Errorf("select event emitter_addrs: %w", err)
+	}
+
+	for rows.Next() {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		var row struct {
+			id          int64
+			emitterAddr []byte
+		}
+		if err := rows.Scan(&row.id, &row.emitterAddr); err != nil {
+			return xerrors.Errorf("read event emitter_addr row: %w", err)
+		}
+		addr, err := address.NewFromBytes(row.emitterAddr)
+		if err != nil {
+			return xerrors.Errorf("parse emitter_addr: %w", err)
+		}
+		emitter, err := actorResolver(ctx, addr, head)
```

Comment:
I'd implement address resolution manually here. Ideally, we'd load the init actor once (https://github.com/filecoin-project/lotus/blob/2e75f3b796dd1396baa37d73ea99dc3407b4da8d/chain/actors/builtin/init/init.go#L32, , then load the address map once (https://github.com/filecoin-project/lotus/blob/2e75f3b796dd1396baa37d73ea99dc3407b4da8d/chain/actors/builtin/init/v13.go#L117), then use that for the rest of this function.

Reply:
@Stebalien I'd love your help on how to keep the index migration tests running after this change. I _think_ I'll have to export `lookupIDFunc` in `StateTree`, unless there is another way you can think of?

Reply:
I think you'd just have to use the `chainStore` directly for this - wrap it in `cbor.NewCborStore`, wrap that in the `AdtStore` like `StateTree#lookupIDinternal` does and proceed from there, `Load`ing the init HAMT and doing a `ResolveAddress` on init state while holding on to the HAMT the whole time. For testing, just build a custom HAMT to hold the addresses you want and pass it in via the `chainStore`. Sounds complicated but would that not work here?

Reply:
I wonder though is there a path to having an address that _doesn't_ resolve to a new ActorID at this point though because the error is a hard fatal and will be annoying for users to encounter.

If there was a revert or two after you shut down your node, upgraded your lotus and tried to do a migration to v4 and encountered an address that failed to resolve because your head tipset is on a diverging branch? You'd be in trouble. Maybe it would be good in this case to error log it but drop the row entirely?

Is there a _right_ thing to do in this case? If we fail to look one up, should we try to load the state from the tipset in the event line and fish around in its init actor?